File: Second Year/2nd Semester/Medii de Proiectare si Programare/Java/Laborator_1/src/main/java/repository/Repository.java
package repository;


import domain.Entity;
import exceptions.InexistentEntityException;

/**
 * CRUD operations repository interface
 * @param <ID> - type E must have an attribute of type ID
 * @param <E> -  type of entities saved in repository
 */
public interface Repository<ID, E extends Entity<ID>> {

    /**
     * Returns the entity that has the given id
     * @param id ID - the id of the entity to be returned
     *           id must not be null
     * @return the entity with the specified id
     * @throws IllegalArgumentException if the id is null.
     * @throws InexistentEntityException if there is no entity with the given id
     */
    E findOne(ID id) throws IllegalArgumentException;

    /**
     * Returns all entities in the repository
     * @return - an iterable object that contains all entities
     */
    Iterable<E> getAll();

    /**
     * Clears the repository's data
     */
    void clear();

    /**
     * Add an entity
     */
    void save(E entity);

    /**
     * Update an entity
     */
    void update(E entity);

    /**
     * Delete an entity
     */
    void delete(ID id);
}

File: Second Year/2nd Semester/Medii de Proiectare si Programare/C#/Laborator_1/repository/Repository.cs
using Laborator_1.domain;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Laborator_1.repository
{
    public interface Repository<ID,E> where E: Entity<ID>
    {
        E FindOne(ID id);
        IEnumerable<E> FindAll();
        void Save(E entity);
        void Delete(ID id);    
        void Update(E entity);
    }
}
